File: functions/feedback.ts
interface Env {
  GITHUB_TOKEN: string;
}

interface FeedbackBody {
  category: string;
  title: string;
  description: string;
  os: string;
  osVersion: string;
  arch: string;
  appVersion: string;
}

const LABEL_MAP: Record<string, string> = {
  bug_report: "bug",
  feature_request: "enhancement",
  general_feedback: "feedback",
  question: "question",
};

const TITLE_PREFIX: Record<string, string> = {
  bug_report: "[Bug Report]",
  feature_request: "[Feature Request]",
  general_feedback: "[Feedback]",
  question: "[Question]",
};

// ── Rate limiting ─────────────────────────────────────────────
// Per-IP sliding window: max 5 submissions per hour.
// Module-level state persists across requests within the same
// worker instance. Good enough for beta-scale traffic.

const MAX_PER_HOUR = 5;
const WINDOW_MS = 60 * 60 * 1000;
const rateLimitMap = new Map<string, number[]>();

function isRateLimited(ip: string): boolean {
  const now = Date.now();
  const timestamps = (rateLimitMap.get(ip) ?? []).filter(t => now - t < WINDOW_MS);
  if (timestamps.length >= MAX_PER_HOUR) return true;
  timestamps.push(now);
  rateLimitMap.set(ip, timestamps);
  return false;
}

// Prevent the map from growing unboundedly if the worker stays
// alive for a long time (e.g. high-traffic edge instance).
function pruneRateLimitMap() {
  const now = Date.now();
  for (const [ip, timestamps] of rateLimitMap) {
    const recent = timestamps.filter(t => now - t < WINDOW_MS);
    if (recent.length === 0) rateLimitMap.delete(ip);
    else rateLimitMap.set(ip, recent);
  }
}

// ── Input limits ──────────────────────────────────────────────
const MAX_TITLE_LEN = 200;
const MAX_DESC_LEN = 10_000;

export const onRequestPost: PagesFunction<Env> = async (context) => {
  // Rate limit by connecting IP (CF-Connecting-IP is always present on CF edge).
  const ip = context.request.headers.get("CF-Connecting-IP") ?? "unknown";
  if (isRateLimited(ip)) {
    // Prune stale entries opportunistically on rejections to keep memory tidy.
    pruneRateLimitMap();
    return Response.json(
      { error: "Too many submissions. Please wait before trying again." },
      { status: 429 },
    );
  }

  let body: FeedbackBody;
  try {
    body = await context.request.json() as FeedbackBody;
  } catch {
    return Response.json({ error: "Invalid JSON" }, { status: 400 });
  }

  const { category, os, osVersion, arch, appVersion } = body;
  let { title, description } = body;

  if (!title?.trim() || !description?.trim()) {
    return Response.json({ error: "Title and description are required" }, { status: 400 });
  }

  // Enforce server-side length caps (client can't be trusted).
  title = title.trim().slice(0, MAX_TITLE_LEN);
  description = description.trim().slice(0, MAX_DESC_LEN);

  const token = context.env.GITHUB_TOKEN;
  if (!token) {
    return Response.json({ error: "Feedback service not configured" }, { status: 503 });
  }

  const prefix = TITLE_PREFIX[category] ?? "[Feedback]";
  const label = LABEL_MAP[category] ?? "feedback";

  const issueBody = [
    "## Description",
    "",
    description,
    "",
    "## System Info",
    "",
    "| Field | Value |",
    "|-------|-------|",
    `| OS | ${os} ${osVersion} |`,
    `| Architecture | ${arch} |`,
    `| App Version | ${appVersion} |`,
    "",
    "---",
    "*Submitted from Harness Kit desktop app*",
  ].join("\n");

  const ghResponse = await fetch(
    "https://api.github.com/repos/harnessprotocol/harness-kit-feedback/issues",
    {
      method: "POST",
      headers: {
        Authorization: `Bearer ${token}`,
        Accept: "application/vnd.github.v3+json",
        "Content-Type": "application/json",
        "User-Agent": "harness-kit-feedback/1.0",
      },
      body: JSON.stringify({
        title: `${prefix} ${title}`,
        body: issueBody,
        labels: [label, "desktop-app"],
      }),
    },
  );

  if (!ghResponse.ok) {
    const err = await ghResponse.text();
    console.error("GitHub API error:", ghResponse.status, err);
    return Response.json({ error: "Failed to create issue" }, { status: 502 });
  }

  const issue = await ghResponse.json() as { html_url: string };
  return Response.json({ issueUrl: issue.html_url });
};

File: apps/desktop/src-tauri/src/commands/feedback.rs
use serde::Serialize;
use std::env;
use std::process::Command as StdCommand;
use tauri::AppHandle;
use tauri_plugin_shell::ShellExt;

// ── Types ────────────────────────────────────────────────────

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GhAuthStatus {
    pub available: bool,
    pub authenticated: bool,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SystemInfo {
    pub os: String,
    pub os_version: String,
    pub arch: String,
    pub app_version: String,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct FeedbackResult {
    pub success: bool,
    pub issue_url: Option<String>,
    pub error: Option<String>,
}

// ── Commands ─────────────────────────────────────────────────

#[tauri::command]
pub async fn check_gh_auth(app: AppHandle) -> Result<GhAuthStatus, String> {
    let shell = app.shell();

    // Check if gh is available
    let version_out = shell.command("gh")
        .args(["--version"])
        .output()
        .await;

    let available = matches!(version_out, Ok(ref o) if o.status.success());
    if !available {
        return Ok(GhAuthStatus { available: false, authenticated: false });
    }

    // Check if authenticated
    let auth_out = shell.command("gh")
        .args(["auth", "status"])
        .output()
        .await
        .map_err(|e| e.to_string())?;

    Ok(GhAuthStatus {
        available: true,
        authenticated: auth_out.status.success(),
    })
}

#[tauri::command]
pub fn get_system_info() -> Result<SystemInfo, String> {
    let os = match env::consts::OS {
        "macos" => "macOS".to_string(),
        "linux" => "Linux".to_string(),
        "windows" => "Windows".to_string(),
        other => other.to_string(),
    };

    let os_version = get_os_version().unwrap_or_else(|| "Unknown".to_string());
    let arch = env::consts::ARCH.to_string();
    let app_version = env!("CARGO_PKG_VERSION").to_string();

    Ok(SystemInfo { os, os_version, arch, app_version })
}

fn get_os_version() -> Option<String> {
    match env::consts::OS {
        "macos" => StdCommand::new("sw_vers")
            .arg("-productVersion")
            .output()
            .ok()
            .filter(|o| o.status.success())
            .map(|o| String::from_utf8_lossy(&o.stdout).trim().to_string()),
        _ => StdCommand::new("uname")
            .arg("-r")
            .output()
            .ok()
            .filter(|o| o.status.success())
            .map(|o| String::from_utf8_lossy(&o.stdout).trim().to_string()),
    }
}

#[tauri::command]
pub async fn submit_feedback(
    app: AppHandle,
    category: String,
    title: String,
    description: String,
    os: String,
    os_version: String,
    arch: String,
    app_version: String,
) -> Result<FeedbackResult, String> {
    let (label, title_prefix) = match category.as_str() {
        "bug_report"       => ("bug",         "[Bug Report]"),
        "feature_request"  => ("enhancement", "[Feature Request]"),
        "general_feedback" => ("feedback",    "[Feedback]"),
        "question"         => ("question",    "[Question]"),
        _                  => ("feedback",    "[Feedback]"),
    };

    let full_title = format!("{} {}", title_prefix, title.trim());

    let body = format!(
        "## Description\n\n{}\n\n## System Info\n\n| Field | Value |\n|-------|-------|\n| OS | {} {} |\n| Architecture | {} |\n| App Version | {} |\n\n---\n*Submitted from Harness Kit desktop app*",
        description.trim(),
        os, os_version,
        arch,
        app_version,
    );

    let shell = app.shell();
    let output = shell.command("gh")
        .args([
            "issue", "create",
            "--repo", "harnessprotocol/harness-kit-feedback",
            "--title", &full_title,
            "--body", &body,
            "--label", label,
            "--label", "desktop-app",
        ])
        .output()
        .await
        .map_err(|e| e.to_string())?;

    if output.status.success() {
        let issue_url = String::from_utf8_lossy(&output.stdout).trim().to_string();
        Ok(FeedbackResult {
            success: true,
            issue_url: Some(issue_url),
            error: None,
        })
    } else {
        let err = String::from_utf8_lossy(&output.stderr).trim().to_string();
        Ok(FeedbackResult {
            success: false,
            issue_url: None,
            error: Some(err),
        })
    }
}
